server/http/test_http/oauth2_complete_test.go

```diff
@@ -0,0 +1,294 @@
+			</body>
+		</html>
+		`,
+			expectedHTTPCode: http.StatusOK,
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			tc.registerResponderFunc()
+
+			tc.handler.OAuth2Complete(tc.w, tc.r)
+
+			assert.Equal(t, tc.expectedHTTPCode, tc.w.StatusCode)
+			assert.Equal(t, tc.expectedHTTPResponse, string(tc.w.Bytes))
+		})
+	}
+}
+
+func badTokenExchangeResponderFunc() {
+	url := "https://login.microsoftonline.com/common/oauth2/v2.0/token"
+
+	responder := httpmock.NewStringResponder(http.StatusBadRequest, `{"error":"invalid request"}`)
+
+	httpmock.RegisterResponder("POST", url, responder)
+}
+
+func unauthorizedTokenGraphAPIResponderFunc() {
+	tokenURL := "https://login.microsoftonline.com/common/oauth2/v2.0/token"
+
+	tokenResponder := httpmock.NewStringResponder(http.StatusOK, `{
+    "token_type": "Bearer",
+    "scope": "user.read%20Fmail.read",
+    "expires_in": 3600,
+    "access_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6Ik5HVEZ2ZEstZnl0aEV1Q...",
+    "refresh_token": "AwABAAAAvPM1KaPlrEqdFSBzjqfTGAMxZGUTdM0t4B4..."
+}`)
+
+	httpmock.RegisterResponder("POST", tokenURL, tokenResponder)
+
+	meRequestURL := "https://graph.microsoft.com/v1.0/me"
+
+	meResponder := httpmock.NewStringResponder(http.StatusUnauthorized, `{
+    "error": {
+        "code": "InvalidAuthenticationToken",
+        "message": "Access token is empty.",
+        "innerError": {
+            "request-id": "d1a6e016-c7c4-4caf-9a7f-2d7079dc05d2",
+            "date": "2019-11-12T00:49:46"
+        }
+    }
+}`)
+
+	httpmock.RegisterResponder("GET", meRequestURL, meResponder)
+}
+
+func statusOKGraphAPIResponderFunc() {
+	tokenURL := "https://login.microsoftonline.com/common/oauth2/v2.0/token"
+
+	tokenResponder := httpmock.NewStringResponder(http.StatusOK, `{
+    "token_type": "Bearer",
+    "scope": "user.read%20Fmail.read",
+    "expires_in": 3600,
+    "access_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6Ik5HVEZ2ZEstZnl0aEV1Q...",
+    "refresh_token": "AwABAAAAvPM1KaPlrEqdFSBzjqfTGAMxZGUTdM0t4B4..."
+}`)
+
+	httpmock.RegisterResponder("POST", tokenURL, tokenResponder)
+
+	meRequestURL := "https://graph.microsoft.com/v1.0/me"
+
+	meResponder := httpmock.NewStringResponder(http.StatusOK, `{
+    "businessPhones": [
+        "businessPhones-value"
+    ],
+    "displayName": "displayName-value",
+    "givenName": "givenName-value",
+    "jobTitle": "jobTitle-value",
+    "mail": "mail-value",
+    "mobilePhone": "mobilePhone-value",
+    "officeLocation": "officeLocation-value",
+    "preferredLanguage": "preferredLanguage-value",
+    "surname": "surname-value",
+    "userPrincipalName": "userPrincipalName-value",
+    "id": "id-value"
+}`)
+
+	httpmock.RegisterResponder("GET", meRequestURL, meResponder)
+}
+
+func getMockBotPoster(ctrl *gomock.Controller, err error) utils.BotPoster {
+	m := mock_utils.NewMockBotPoster(ctrl)
+	m.EXPECT().PostDirect(gomock.Any(), gomock.Any(), gomock.Any()).Return(err).AnyTimes()
```

Comment:
I think putting the mock Expects into a function like this seems good at first, but then I realized that you're not actually testing whether the post was called or not (because of `AnyTimes`). I think that goes for all the mocks. You might have to redesign the tests a bit so that you can specify exactly what each mock is called with (to verify that they are being called with the expected arguments -- right now you're using `gomock.Any()`) and how many times they were called (so that the tests error out when the mock is /not/ called, or called more than you expected). Right now, the botPoster could be called zero times, and the test would still pass, right?

Reply:
That is correct. When I was writing those functions it did feel like a bit of a cheat to just accept anything and just let it get calls any number of times. I have figured out a solution to build the mocks out with expected parameters and max number of times for each call. I will be pushing that soon.